### PR TITLE
Do not commit subframe navigations for ipfs

### PR DESCRIPTION
**Pull request by cypt4**
Fixes https://github.com/brave/brave-browser/issues/24211

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: patches/content-browser-renderer_host-navigation_request.cc.patch [hunk 1]
```diff
@@ -18,3 +18,11 @@ index e3e73a2353312a505ffee9d7b9a801bf480dbf86..cab6855d7eab85ced7cd482c27248c2c
    // TODO(clamy): Avoid cloning the navigation params and create the
    // ResourceRequest directly here.
    std::vector<std::unique_ptr<NavigationLoaderInterceptor>> interceptor;
+@@ -7525,6 +7527,7 @@ bool NavigationRequest::MaybeCancelFailedNavigation() {
+   // account as well.
+   // If the request was canceled by the user, do not show an error page.
+   if (net::ERR_ABORTED == net_error_ ||
++      BRAVE_ONIPFSIPNS_MAYBE_CANCEL_NAVIGATION
```
Comment (iefremov) [inline patches/content-browser-renderer_host-navigation_request.cc.patch:25]: can't we do that via NavigationThrottle ?

File: browser/ipfs/ipfs_subframe_navigation_throttle.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      content::NavigationHandle* navigation_handle);
+  ~IpfsSubframeNavigationThrottle() override;
+
+  IpfsSubframeNavigationThrottle(const IpfsSubframeNavigationThrottle&) =
+      delete;
+  IpfsSubframeNavigationThrottle& operator=(
+      const IpfsSubframeNavigationThrottle&) = delete;
+
+  // content::NavigationThrottle implementation:
+  content::NavigationThrottle::ThrottleCheckResult WillFailRequest() override;
```
Comment (iefremov) [inline browser/ipfs/ipfs_subframe_navigation_throttle.h:31]: add comments pls

File: browser/ipfs/ipfs_subframe_navigation_throttle.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+#define BRAVE_BROWSER_IPFS_IPFS_SUBFRAME_NAVIGATION_THROTTLE_H_
+
+#include <memory>
+
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/navigation_throttle.h"
+
+namespace ipfs {
+
+class IpfsSubframeNavigationThrottle : public content::NavigationThrottle {
```
Comment (iefremov) [inline browser/ipfs/ipfs_subframe_navigation_throttle.h:16]: class-level comment

**Review by iefremov — APPROVED** [review]

**Review by yrliou — APPROVED** [review]

**Comment by stephendonner** [comment]
## Verified `PASSED` using 

Brave | 1.44.21 Chromium: 104.0.5112.81 (Official Build) nightly (x86_64)
-- | --
Revision | 5b7b76419d50f583022568b6764b630f6ddc9208-refs/branch-heads/5112@{#1309}
OS | macOS Version 13.0 (Build 22A5311f)

Confirmed the crash using `1.42.86`, and the privately-provided minimized testcase.

Worked with @cypt4 and used the minimal testcases from both this issue and these others, collected from HackerOne in https://github.com/brave/brave-browser/issues/24093.

minimized testcase | testcase 1 | testcase 2 | testcase 3 | testcase 4 | testcase 5 | testcase 6
---------------------|----------|------------|------------|-----------|------------|-----------
Screenshot 2022-08-04 at 3 33 10 PM | Screenshot 2022-08-04 at 3 29 13 PM | Screenshot 2022-08-04 at 3 29 20 PM | Screenshot 2022-08-04 at 3 29 28 PM | Screenshot 2022-08-04 at 3 29 37 PM | Screenshot 2022-08-04 at 3 29 44 PM | Screenshot 2022-08-04 at 3 29 50 PM

**Comment by stephendonner** [comment]
@cypt4 @spylogsster can we get an uplift for this started?  Am assuming we'll want it at least in 1.43.x, if not also 1.42.x, via a hotfix/maintenance release?

**Pull request merged**